File: next-frontend/api/definitions/models/feed.ts
import { CategoryWithIdSchema } from "@/api/definitions/models/category";
import { TagWithIdSchema } from "@/api/definitions/models/tag";
import { z } from "zod";

export const EventTypeSchema = z.enum([
  "session_completed",
  "session_started",
  "task_completed",
  "project_completed",
]);

export const ReadUserAvatarSchema = z.object({
  avatar_url: z.string().nullable(),
  id: z.string(),
  username: z.string(),
});

export const ReadFeedReactionSchema = z.object({
  created_at: z.coerce.date<Date>(),
  emoji: z.string(),
  id: z.uuid(),
  user: ReadUserAvatarSchema,
});

export const FeedSessionProjectSchema = z.object({
  color: z.string(),
  id: z.string().uuid(),
  name: z.string(),
});

export const FeedSessionTaskSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
});

export const SessionCompletedEventSchema = z.object({
  category: CategoryWithIdSchema.omit({ last_used_at: true }),
  description: z.string().nullish(),
  end_time: z.coerce.date<Date>(),
  project: FeedSessionProjectSchema.nullable().optional(),
  session_id: z.string(),
  start_time: z.coerce.date<Date>(),
  tags: z.array(TagWithIdSchema),
  task: FeedSessionTaskSchema.nullable().optional(),
});

export const FeedProjectSchema = z.object({
  color: z.string(),
  id: z.string().uuid(),
  image_url: z.string().nullable(),
  name: z.string(),
});

export const TaskCompletedEventSchema = z.object({
  hours_of_work: z.number(),
  project: FeedProjectSchema,
  task_description: z.string().nullable(),
  task_id: z.string().uuid(),
  task_name: z.string(),
});

export const TaskTimeBreakdownSchema = z.object({
  minutes: z.number(),
  task_id: z.string().uuid(),
  task_name: z.string(),
});

export const CategoryTimeBreakdownSchema = z.object({
  category_color: z.string(),
  category_id: z.string().uuid(),
  category_name: z.string(),
  minutes: z.number(),
});

export const ProjectCompletedEventSchema = z.object({
  categories_time_breakdown: z.array(CategoryTimeBreakdownSchema),
  project_color: z.string(),
  project_description: z.string().nullable(),
  project_id: z.string().uuid(),
  project_image_url: z.string().nullable(),
  project_name: z.string(),
  tasks_time_breakdown: z.array(TaskTimeBreakdownSchema),
});

export const SourceTypeMappingSchema = z.discriminatedUnion("source_type", [
  z.object({
    source_data: ReadUserAvatarSchema,
    source_type: z.literal("user"),
  }),
  z.object({
    source_data: z.number(),
    source_type: z.literal("placeholder"),
  }),
]);

export const EventTypeMappingSchema = z.discriminatedUnion("event_type", [
  z.object({
    event_data: SessionCompletedEventSchema,
    event_type: z.literal("session_completed"),
  }),
  z.object({
    event_data: TaskCompletedEventSchema,
    event_type: z.literal("task_completed"),
  }),
  z.object({
    event_data: ProjectCompletedEventSchema,
    event_type: z.literal("project_completed"),
  }),
]);

export const ReadFeedEventSchema = z
  .object({
    created_at: z.coerce.date<Date>(),
    id: z.uuid(),
    reactions: z.array(ReadFeedReactionSchema),
  })
  .and(EventTypeMappingSchema.and(SourceTypeMappingSchema));

export const ReadFeedSubscriptionSchema = z
  .object({
    created_at: z.coerce.date<Date>(),
    id: z.uuid(),
    is_muted: z.boolean(),
    is_paused: z.boolean(),
  })
  .and(SourceTypeMappingSchema);

export const UpdateFeedSubscriptionSchema = z.object({
  is_muted: z.boolean().optional(),
  is_paused: z.boolean().optional(),
  subscription_id: z.uuid(),
});

export type EventTypeMapping = z.infer<typeof EventTypeMappingSchema>;
export type FeedEventType = z.infer<typeof EventTypeSchema>;
export type ReadFeedEvent = z.infer<typeof ReadFeedEventSchema>;
export type ReadFeedReaction = z.infer<typeof ReadFeedReactionSchema>;
export type ReadFeedSubscription = z.infer<typeof ReadFeedSubscriptionSchema>;
export type ReadUserAvatar = z.infer<typeof ReadUserAvatarSchema>;
export type UpdateFeedSubscription = z.infer<
  typeof UpdateFeedSubscriptionSchema
>;

export const FeedSourceTypeSchema = z.union([
  z.literal("user"),
  z.literal("placeholder"),
]);
export type FeedSourceType = z.infer<typeof FeedSourceTypeSchema>;

File: backend/src/entity/feed.rs
use chrono::{DateTime, Local};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::dto::user::read_user::ReadUserDto;
use crate::entity::category::Category;
use crate::entity::tag::Tag;

#[derive(Clone, Serialize, Deserialize, Debug)]
#[serde(tag = "event_type", content = "event_data", rename_all = "snake_case")]
pub enum FeedEventType {
    SessionCompleted(SessionEventData),
    TaskCompleted(TaskEventData),
    ProjectCompleted(ProjectEventData),
}

#[derive(Clone, Serialize, Deserialize, Debug)]
#[serde(
    tag = "source_type",
    content = "source_data",
    rename_all = "snake_case"
)]
pub enum FeedEventSource {
    User(ReadUserDto),
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct FeedEvent {
    pub id: Uuid,
    #[serde(flatten)]
    pub source: FeedEventSource,
    #[serde(flatten)]
    pub data: FeedEventType,
    pub created_at: DateTime<Local>,
}

// possible events
#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct SessionEventData {
    pub session_id: Uuid,
    pub category: FeedSessionCategory,
    pub tags: Vec<FeedSessionTag>,
    pub description: Option<String>,
    pub start_time: DateTime<Local>,
    pub end_time: DateTime<Local>,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct TaskEventData {
    pub task_id: Uuid,
    pub task_name: String,
    pub task_description: Option<String>,
    pub hours_of_work: f64,
    pub project: FeedProject,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct ProjectEventData {
    pub project_id: Uuid,
    pub project_name: String,
    pub project_description: Option<String>,
    pub project_color: String,
    pub project_image_url: Option<String>,
    pub tasks_time_breakdown: Vec<TaskTimeBreakdown>,
    pub categories_time_breakdown: Vec<CategoryTimeBreakdown>,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct TaskTimeBreakdown {
    pub task_id: Uuid,
    pub task_name: String,
    pub minutes: f64,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct CategoryTimeBreakdown {
    pub category_id: Uuid,
    pub category_name: String,
    pub category_color: String,
    pub minutes: f64,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct FeedReaction {
    pub id: Uuid,
    pub feed_event_id: Uuid,
    pub user: ReadUserDto,
    pub emoji: String,
    pub created_at: DateTime<Local>,
}

// INFO: mock objects to prevent growing dependencies
#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct FeedSessionCategory {
    pub id: Uuid,
    pub name: String,
    pub color: String,
}

// INFO: mock objects to prevent growing dependencies
#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct FeedSessionTag {
    pub id: Uuid,
    pub label: String,
    pub color: String,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct FeedProject {
    pub id: Uuid,
    pub name: String,
    pub color: String,
    pub image_url: Option<String>,
}

impl From<Tag> for FeedSessionTag {
    fn from(tag: Tag) -> Self {
        Self {
            id: tag.id,
            label: tag.label,
            color: tag.color,
        }
    }
}
impl From<Category> for FeedSessionCategory {
    fn from(category: Category) -> Self {
        Self {
            id: category.id,
            name: category.name,
            color: category.color,
        }
    }
}
